File: TCP  IP Server Client Example Part 7 (Chat Application)/[C#][VB]-TCP  IP Server Client Example Part 7 (Chat Application)/C# and VB/ServerClientChat1.3/Classes/PublicIP.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;

namespace ServerClientChat1._3.Classes
{
    class PublicIP
    {

        public delegate void PublicIPKnownHandler(string PublicIP);
        public event PublicIPKnownHandler PublicIPKnown;
        public delegate void PublicIPErrorHandler(string Message);
        public event PublicIPErrorHandler PublicIPError;

        private int RetryAttempts = 0;

        public void GetPublicIpAddress()
        {
         
                Task tGetIP = new Task(() => GetPublicIpAddressAsync());
                tGetIP.Start();
           
        }

        /// <summary>
        /// Contact the ifConfig.Me webservice to get our public IP Address
        /// </summary>
        /// <returns>
        /// String indicating our public IP Address
        /// </returns>
        private void GetPublicIpAddressAsync()
        {
            try
            {
                var request = (HttpWebRequest)WebRequest.Create("http://ifconfig.me");

                request.UserAgent = "curl"; // this simulate curl linux command

                string publicIPAddress;

                request.Method = "GET";
                using (WebResponse response = request.GetResponse())
                {
                    using (var reader = new StreamReader(response.GetResponseStream()))
                    {
                        publicIPAddress = reader.ReadToEnd();
                    }
                }

                PublicIPKnown(publicIPAddress.Replace("\n", ""));
            }
            catch (Exception ex)
            {
                if (RetryAttempts < 4) // If we just keep calling ourselves without any limit we will get a StackOverFlow Error
                {
                    GetPublicIpAddressAsync();
                    RetryAttempts++;
                }
                else
                {
                    PublicIPError(ex.Message);
                }
            }
           
            
        }

    }
}

File: TCP  IP Server Client Example Part 7 (Chat Application)/[VB]-TCP  IP Server Client Example Part 7 (Chat Application)/VB/ServerClientChat1.3/Classes/PublicIP.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Net
Imports System.Text
Imports System.Threading.Tasks

Namespace ServerClientChat1._3.Classes
	Friend Class PublicIP

		Public Delegate Sub PublicIPKnownHandler(ByVal PublicIP As String)
		Public Event PublicIPKnown As PublicIPKnownHandler
		Public Delegate Sub PublicIPErrorHandler(ByVal Message As String)
		Public Event PublicIPError As PublicIPErrorHandler

		Private RetryAttempts As Integer = 0

		Public Sub GetPublicIpAddress()

				Dim tGetIP As New Task(Sub() GetPublicIpAddressAsync())
				tGetIP.Start()

		End Sub

		''' <summary>
		''' Contact the ifConfig.Me webservice to get our public IP Address
		''' </summary>
		''' <returns>
		''' String indicating our public IP Address
		''' </returns>
		Private Sub GetPublicIpAddressAsync()
			Try
				Dim request = CType(WebRequest.Create("http://ifconfig.me"), HttpWebRequest)

				request.UserAgent = "curl" ' this simulate curl linux command

				Dim publicIPAddress As String

				request.Method = "GET"
				Using response As WebResponse = request.GetResponse()
					Using reader = New StreamReader(response.GetResponseStream())
						publicIPAddress = reader.ReadToEnd()
					End Using
				End Using

				RaiseEvent PublicIPKnown(publicIPAddress.Replace(vbLf, ""))
			Catch ex As Exception
				If RetryAttempts < 4 Then ' If we just keep calling ourselves without any limit we will get a StackOverFlow Error
					GetPublicIpAddressAsync()
					RetryAttempts += 1
				Else
					RaiseEvent PublicIPError(ex.Message)
				End If
			End Try


		End Sub

	End Class
End Namespace
